[h-b-k-nishi] iP

h-b-k-nishi opened the pull request:
# Chatbot Pochi

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Pochi frees your mind of having to remember things you need to do. It's,

- text-based
- ~~EASY~~ *SUPER* EASY to learn
- ~~FAST~~ *SUPER* FAST to use

All you need to do is,

1. download it from [here](https://github.com/h-b-k-nishi/ip/releases/tag/A-Jar).
1. double-click it.
1. add your tasks.
1. let it manage your tasks for you :wink:

And it is **FREE**!

Features:

- [x] Managing tasks
- [x] Managing deadlines and events
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice `Java` too. Here's the main method:

```Java
public class Main {
    public static void main(String[] args) {
        Application.launch(Pochi.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Commit:
Rename chatbot to Pochi

Commit:
Add javadoc

Commit:
Implement echo feature

Commit:
Create Storage.java and Task.java

Commit:
Create an instance method to isolate I/O procedure from main

Commit:
Implement add and list features

Commit:
Change the name of field from name to description

Commit:
Implement mark / unmark functionality

Commit:
Implement Todo, Deadline, and Event

Commit:
Implement a text UI testing

Commit:
Put @Override annotations

Commit:
Implement all the necessary exceptions

Commit:
Close the scanner at the end of whole process

Commit:
Remove unnecessary change of line

Commit:
Implement the delete functionality

Commit:
Change the names of constants to capital

Commit:
Add JavaDoc for Task::of

Commit:
Implement a feature to create and auto-save a log of tasks

Commit:
Merge branch 'branch-Level-7'

Commit:
Removed unnecessary import

Commit:
Have Pochi recognize dates and times

Commit:
Merge branch 'branch-Level-8'

Commit:
Organize classes based on OOP principle

Commit:
Reconstruct some classes based on OOP principles

Commit:
Implement a feature to save and load the log from the previous session

Commit:
Packaged all the classes

Commit:
Remove redundant files

Commit:
Update input.txt

Commit:
Merge branch 'branch-A-Gradle'

Commit:
Modify setting of gradle

Commit:
Implement test classes

Commit:
Merge branch 'branch-A-JUnit'

Commit:
Remove unnecessary comment

Commit:
Rename the test classes so that gradle can recognize them

Commit:
Remove unnecessary files

Commit:
Fix java docs

Commit:
Mofify names to comply with coding standard

Commit:
Revise indents to comply with coding standard

Commit:
Reorder the import statements

Commit:
Fix java doc to comply with the coding standard

Commit:
Implement the search-by-keyword feature

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Resolve the merge conflict

Commit:
Resove the merge conflict

Commit:
Rewrite some codes to abide by the coding standard

Commit:
Enable the checkstyle feature with gradle

Commit:
Change a code a bit to commit some modification

Commit:
Merge branch 'branch-A-CheckStyle'

Commit:
Update build.gradle as pre-required

Commit:
Implement GUI up to the Tutorial 3

Commit:
Develop a basic UI up to the tutorial 4

Commit:
Migrate from CUI to GUI

Commit:
Adjust the size of text and icon

Commit:
Enable reading of past logs in the GUI environment

Commit:
Merge branch 'branch-Level-10'

TasmiaH0508 commented on src/main/java/pochi/javafx/DialogBox.java:
```diff
@@ -0,0 +1,71 @@
+        displayPicture.setImage(icon);
+    }
+
+    private void flip() {
```
I think you may have accidentally missed out JavaDoc for this function.

ryanlow64 replied:
its a private helper

TasmiaH0508 commented on src/main/java/pochi/core/Parser.java:
```diff
@@ -0,0 +1,116 @@
+ */
+public class Parser {
+
+    private static List<String> parseDescriptions(List<String> descriptions, List<String> separators) {
```
I think you may have accidentally missed out JavaDoc here.

ryanlow64 replied:
I think its fine without JavaDoc. Its private static and the name says what it does already.

TasmiaH0508 commented on src/main/java/pochi/core/Pochi.java:
```diff
@@ -0,0 +1,136 @@
+    }
+
+    /**
+     * Responses to input commands from the user.
```
Perhaps changing the description to : "Returns a response as a string to input commands from the user. " would make the description clearer in the sense that the messages are not printed in the console but instead a message string is returned.

TasmiaH0508 reviewed (commented):
Hello, I was reviewing your code for coding standard related issues. 

Overall, your code has little to no violations and you did a great job:)

ryanlow64 commented on src/main/java/pochi/core/Parser.java:
```diff
@@ -0,0 +1,116 @@
+        List<String> commands = List.of(command.split(" "));
+
+        if (commands.isEmpty()) {
+            throw new EmptyCommandException();
```
good use of your own exceptions for guard statements

ryanlow64 commented on src/main/java/pochi/core/Parser.java:
```diff
@@ -0,0 +1,116 @@
+
+        results.add(commands.get(0));
+
+        if (commands.get(0).equals("bye") || commands.get(0).equals("list")) {
```
you can store commands.get(0) somewhere since you already checked whether its empty in line 67

ryanlow64 commented on src/main/java/pochi/core/Pochi.java:
```diff
@@ -0,0 +1,136 @@
+    private String processCommand(List<String> commands) throws CommandException {
+        String res = "";
+        if (commands.get(0).equals("list")) {
+            res += tasks.getStatus() + "\n";
```
try to use `System.lineSeparator()` because different OS use different characters (e.g. `\r`, `\r\n`) for linebreaks. Might cause a problem in tP.

ryanlow64 commented on src/main/java/pochi/core/TaskList.java:
Your logic for guard statements is flipped. Usually it makes more sense to do input checking first, then proceed with the function. (See an e.g. below)

ryanlow64 commented on src/main/java/pochi/core/TaskList.java:
```diff
@@ -0,0 +1,157 @@
+        } else {
+            throw new IndexOutOfBoundException(tasks.size());
+        }
+    }
```
```suggestion
    public Task deleteTask(int index) throws IndexOutOfBoundException {
        if (index < 1 || index > tasks.size()) {
            throw new IndexOutOfBoundException(tasks.size());
        }
        return tasks.remove(index - 1);
    }
```

ryanlow64 commented on src/main/java/pochi/tasks/Deadline.java:
```diff
@@ -0,0 +1,49 @@
+     */
+    public Deadline(String description, LocalDateTime deadline) throws EmptyDescriptionException {
+        super(description);
+
```
```suggestion

```

ryanlow64 replied:
I think whitespace is only necessary for separating code blocks with different logic. (Just my own opinion)

ryanlow64 commented on src/test/java/pochi/core/TaskTest.java:
```diff
@@ -0,0 +1,82 @@
+     */
+    @Test
+    public void missingArgumentTest() {
+        try {
+            Task task = Task.createTask(List.of("todo", "return book"));
+
+            assertEquals(0, 1);
+        } catch (Exception e) {
+            assertEquals(e.getMessage(),
+                    "Some arguments (/by, /from, /to, or the index of task) are missing!!");
+        }
```
```suggestion
        assertThrows(MissingArgumentException.class, () -> Task.createTask(List.of("todo", "return book")));
```

ryanlow64 commented on src/test/java/pochi/core/TaskTest.java:
Other than `assertEquals`, you can assert many things.
e.g. `assertThrows`, `assertTrue`, `assertInstanceOf`

ryanlow64 reviewed (commented):
Well done, no major crisis to me.
Some improvements can be made to make your code more robust

Commit:
Added assertions for parameter validation

Ui::formatDateTime assumes the given dateAndTime parameter is not null,
and Task::createTask assumes the given descriptions parameter is
neither empty nor contains invalid commands
other than todo, deadline, or event.

Since there is a possibility of these exceptional cases occurring,
it is better to add assertions for validation.

Add assertions that verifies
* The dateAndTime parameter passed to Ui::formatDateTime is not null.
* The descriptions parameter passed to Task::createTask is not empty.
* The descriptions parameter passed to Task::createTask has
“todo”, “deadline”, or “event” as its first parameter.

Assertions are preferred over exceptions in this case, as these errors
are likely caused by programmer mistakes rather than user input errors.

Commit:
Add assertions for parameter validation

Ui::formatDateTime assumes the given dateAndTime parameter is not null,
and Task::createTask assumes the given descriptions parameter is
neither empty nor contains invalid commands
other than todo, deadline, or event.

Since there is a possibility of these exceptional cases occurring,
it is better to add assertions for validation.

Add assertions that verifies
* The dateAndTime parameter passed to Ui::formatDateTime is not null.
* The descriptions parameter passed to Task::createTask is not empty.
* The descriptions parameter passed to Task::createTask has
“todo”, “deadline”, or “event” as its first parameter.

Assertions are preferred over exceptions in this case, as these errors
are likely caused by programmer mistakes rather than user input errors.

Commit:
Enhance code quality with minor modifications

Commit:
Enhance code quality with step-by-step refinements

The current code suffers from poor readability and organization.

To improve future maintainability, the code structure needs to be
cleaned up and made more readable.

* Remove all redundant magic literals.
* Restructure functions for better clarity.

Magic literals can lead to subtle, hard-to-detect bugs such as typos,
so eliminating them improves code reliability.

h-b-k-nishi committed:
Merge pull request #1 from h-b-k-nishi/branch-A-Assertions

A-Assertions

Commit:
Resolve merge conflicts with the current master

h-b-k-nishi committed:
Merge pull request #2 from h-b-k-nishi/branch-A-CodeQuality

A-CodeQuality

Commit:
Allow the user to specify a date by day

Commit:
Set the title and align the user input to right

Commit:
Merge branch 'branch-A-BetterGui'

Commit:
Add the complete version of user guide

Commit:
Enclose some keywords in backticks

Commit:
Add javafx plugin to build.gradle

Commit:
Fix a bug in file creation

Commit:
Implement a feature to exit with 'bye'

Commit:
Fix a bug when receiving a non-integer index
